examples/scripts/learning_regularization_parameter_maps_for_tv_reconstruction.py

```diff
@@ -0,0 +1,876 @@
+            recon_matrix=self.img_shape,
+            encoding_matrix=self.k_shape,
+            traj=traj,
+        )
+        csm_operator = mrpro.operators.SensitivityOp(csm)
+        forward_operator = fourier_operator @ csm_operator
+        stacked_operator = mrpro.operators.LinearOperatorMatrix(((forward_operator,), (self.gradient_operator,)))
+
+        primal_stepsize = dual_stepsize = 0.95 * 1.0 / self.stacked_operator_norm
+
+        (solution,) = mrpro.algorithms.optimizers.pdhg(
+            f=f,
+            g=self.g,
+            operator=stacked_operator,
+            initial_values=(initial_image,),
+            max_iterations=self.n_iterations,
+            primal_stepsize=primal_stepsize,
+            dual_stepsize=dual_stepsize,
+            tolerance=0.0,
+        )
+        return solution
+
+
+# %% [markdown]
+# ### Creating the training data
+# In the following, we create some training data for the network. We use some images borrowed from the
+# BrainWeb dataset [[Aubert-Broche et al, IEEE TMI 2006](https://ieeexplore.ieee.org/abstract/document/1717639)],
+# which is a simulated MRI dataset and for which MRpro provides a simple interface to load the data.
+# For simplicity, we here use some images that were generated by simulating the contrast using the
+# inversion recovery signl model in MRpro.
+# First, we start by defining some auxiliary functions that we will need for retrospectively simulating undersampled
+# and noisy k-space data.
+
+# %%
+
+
+def add_gaussian_noise(kdata: torch.Tensor, noise_variance: float, seed: int) -> torch.Tensor:
+    """Corrupt given k-space data with additive Gaussian noise."""
+    rng = torch.Generator().manual_seed(seed)
+    kdata = kdata + noise_variance * kdata.std() * torch.randn(
+        kdata.shape, dtype=kdata.dtype, device=kdata.device, generator=rng
+    )
+
+    return kdata
+
+
+def normalize_kspace_data_and_image(
+    kdata: torch.Tensor, target_image: torch.Tensor | None, factor: float = 100
+) -> tuple[torch.Tensor, torch.Tensor] | torch.Tensor:
+    """
+    Normalize k-space data and (possibly) target image.
+
+    Divide by std multiply by (an empirically set) factor.
+    """
+    kdata_norm = torch.linalg.norm(kdata)
+
+    kdata /= kdata_norm
+    kdata *= factor
+    if target_image is not None:
+        target_image /= kdata_norm
+        target_image *= factor
+
+        return kdata, target_image
+    else:
+        return kdata
+
+
+# %% [markdown]
+# Further, we define a torch dataset that will be used for training the network.
+# %%
+
+
+class Cartesian2D(torch.utils.data.Dataset):
+    r"""Dataset for 2D problems."""
+
+    def __init__(self, images: torch.Tensor) -> None:
+        """Parameters
+
+        ----------
+        images
+            images to be used in the dataset.
+            Should have the shape (others, coils, z, y, x) with z=1.
+        """
+        self.images = images
+
+    def prepare_data(
```

Comment:
Why is this an instance method of a dataset?
Please refactor as already discussed in length.

Reply:
Either have a __getitem__ that returns target image, kdata, fourierop with collate funciton for fourier op,
or a tensordataset of images and have a prepare data function or a CartesianProblem class with
n_coils etc set in init and a __call__ that is imagetensor -> (kdata, aquisitionoperator)

and the forward of the network be (kdata, aquisitionoperator) -> image